File: apps/server/AliasVault.Client/Services/VaultSync/VaultKeyService.cs
//-----------------------------------------------------------------------
// <copyright file="VaultKeyService.cs" company="aliasvault">
// Copyright (c) aliasvault. All rights reserved.
// Licensed under the AGPLv3 license. See LICENSE.md file in the project root for full license information.
// </copyright>
//-----------------------------------------------------------------------

namespace AliasVault.Client.Services.VaultSync;

using System.Net.Http.Json;
using AliasVault.Client.Services.JsInterop;
using AliasVault.Client.Services.JsInterop.RustCore;
using AliasVault.Client.Services.VaultSync.Exceptions;
using AliasVault.Client.Services.VaultSync.Models;
using AliasVault.Shared.Models.Enums;
using AliasVault.Shared.Models.WebApi.V1.Auth;
using AliasVault.Shared.Models.WebApi.V2.Auth;
using Blazored.LocalStorage;
using Microsoft.JSInterop;

/// <summary>
/// Fetches, decrypts and caches the account-key unlock chain.
/// </summary>
/// <param name="httpClient">The HTTP client.</param>
/// <param name="localStorage">The local storage service.</param>
/// <param name="jsInteropService">JsInteropService instance.</param>
/// <param name="rustCoreService">RustCoreService instance.</param>
/// <param name="logger">ILogger instance.</param>
public sealed class VaultKeyService(HttpClient httpClient, ILocalStorageService localStorage, JsInteropService jsInteropService, RustCoreService rustCoreService, ILogger<VaultKeyService> logger)
{
    private static readonly string VaultKeyEndpoint = ApiRoute($"VaultKey/{UnlockMethodTypes.ToToken(UnlockMethodType.Password)}");

    /// <summary>
    /// Fetch the current user's password vault key from the server.
    /// </summary>
    /// <returns>The vault key, or null when the account has no key chain (legacy account).</returns>
    public async Task<VaultKeyResponse?> FetchVaultKeyAsync()
    {
        var body = await httpClient.GetFromJsonAsync<VaultKeyGetResponse>(VaultKeyEndpoint);
        return body?.VaultKey;
    }

    /// <summary>
    /// Resolve the vault encryption key right after authentication: fetch the vault key from the server, decrypt the
    /// Account Key with the derived key (KEK), and with the AK decrypt the VEK and the account private key. All
    /// encrypted blobs are cached locally. For a legacy account (server explicitly reports no vault key) the derived
    /// key itself is the encryption key and any stale cached chain is cleared.
    /// </summary>
    /// <param name="derivedKeyBase64">The password-derived key.</param>
    /// <returns>The resolved keys.</returns>
    /// <exception cref="VaultKeyDecryptException">Thrown when the chain does not open with the derived key.</exception>
    public async Task<ResolvedVaultKey> ResolveEncryptionKeyAsync(string derivedKeyBase64)
    {
        var vaultKey = await FetchVaultKeyAsync();
        if (vaultKey is null)
        {
            await ClearCachedChainAsync();
            return new ResolvedVaultKey(derivedKeyBase64, null, true);
        }

        var resolved = await DecryptKeyChainAsync(vaultKey.EncryptedAccountKey, vaultKey.EncryptedVek, vaultKey.EncryptedAccountPrivateKey, derivedKeyBase64);
        await CacheVaultKeyBlobsAsync(vaultKey);
        return resolved;
    }

    /// <summary>
    /// Unlock with the master password: fetch the chain when online (its own derivation parameters decide the KEK),
    /// fall back to the cached chain and parameters when the server cannot be reached.
    /// </summary>
    /// <param name="username">The username, needed to look up the salt of a legacy account.</param>
    /// <param name="password">The master password.</param>
    /// <returns>The resolved keys. A legacy result still has to be validated by the caller.</returns>
    /// <exception cref="VaultKeyDecryptException">Thrown when the password is wrong.</exception>
    /// <exception cref="VaultKeyUnavailableException">Thrown when offline and the chain this account unlocks with is not cached.</exception>
    public async Task<ResolvedVaultKey> UnlockWithPasswordAsync(string username, string password)
    {
        VaultKeyResponse? vaultKey = null;
        var online = false;
        Exception? fetchError = null;
        try
        {
            vaultKey = await FetchVaultKeyAsync();
            online = true;
        }
        catch (Exception ex) when (ex is HttpRequestException { StatusCode: null } or TaskCanceledException)
        {
            // No response at all (network down, DNS, timeout): attempt an offline unlock. A server error response propagates.
            logger.LogWarning(ex, "Vault key could not be fetched, attempting offline unlock.");
            fetchError = ex;
        }

        if (vaultKey is not null)
        {
            var kek = await DeriveKeyAsync(password, vaultKey.Salt, vaultKey.EncryptionSettings);
            var resolved = await DecryptKeyChainAsync(vaultKey.EncryptedAccountKey, vaultKey.EncryptedVek, vaultKey.EncryptedAccountPrivateKey, kek);
            await CacheVaultKeyBlobsAsync(vaultKey);
            return resolved;
        }

        if (online)
        {
            // Legacy account: the derived key is the encryption key and the salt comes from the login handshake.
            var initiate = await httpClient.PostAsJsonAsync(ApiRoute("Auth/login"), new LoginInitiateRequest(username));
            initiate.EnsureSuccessStatusCode();
            var loginResponse = await initiate.Content.ReadFromJsonAsync<LoginInitiateResponse>() ?? throw new InvalidOperationException("Empty login initiate response.");
            await StoreDerivationParamsAsync(new EncryptionKeyDerivationParams(loginResponse.Salt, loginResponse.EncryptionType, loginResponse.EncryptionSettings));
            await ClearCachedChainAsync();
            return new ResolvedVaultKey(await DeriveKeyAsync(password, loginResponse.Salt, loginResponse.EncryptionSettings), null, true);
        }

        // Offline: derive from the cached parameters and open the cached chain.
        var parameters = await GetDerivationParamsAsync() ?? throw new VaultKeyUnavailableException(fetchError);
        var offlineKek = await DeriveKeyAsync(password, parameters.Salt, parameters.EncryptionSettings);
        var cached = await ResolveFromLocalCacheAsync(offlineKek);
        if (cached is not null)
        {
            return cached;
        }

        return new ResolvedVaultKey(offlineKek, null, true);
    }

    /// <summary>
    /// Whether the given derived key opens the cached chain.
    /// </summary>
    /// <param name="derivedKeyBase64">The password-derived key to test.</param>
    /// <returns>True or false when a chain is cached, null when there is none to test against.</returns>
    public async Task<bool?> TryOpenCachedChainAsync(string derivedKeyBase64)
    {
        var encryptedAccountKey = await localStorage.GetItemAsStringAsync(StorageKeys.EncryptedAccountKey);
        if (string.IsNullOrEmpty(encryptedAccountKey))
        {
            return null;
        }

        try
        {
            await DecryptKeyOrThrowAsync(encryptedAccountKey, derivedKeyBase64);
            return true;
        }
        catch (VaultKeyDecryptException)
        {
            return false;
        }
    }

    /// <summary>
    /// The cached account public key, or null when the account has no keypair yet.
    /// </summary>
    /// <returns>The public key as a JWK JSON string.</returns>
    public async Task<string?> GetAccountPublicKeyAsync()
    {
        return await localStorage.GetItemAsStringAsync(StorageKeys.AccountPublicKey);
    }

    /// <summary>
    /// Persist the KEK derivation parameters.
    /// </summary>
    /// <param name="parameters">The parameters.</param>
    /// <returns>Task.</returns>
    public async Task StoreDerivationParamsAsync(EncryptionKeyDerivationParams parameters)
    {
        await localStorage.SetItemAsync(StorageKeys.EncryptionKeyDerivationParams, parameters);
    }

    /// <summary>
    /// The cached KEK derivation parameters, or null when none are cached.
    /// </summary>
    /// <returns>The parameters.</returns>
    public async Task<EncryptionKeyDerivationParams?> GetDerivationParamsAsync()
    {
        try
        {
            return await localStorage.GetItemAsync<EncryptionKeyDerivationParams>(StorageKeys.EncryptionKeyDerivationParams);
        }
        catch (Exception ex)
        {
            logger.LogWarning(ex, "Cached key derivation parameters are unreadable, ignoring them.");
            return null;
        }
    }

    /// <summary>
    /// Persist a server vault-key response's encrypted blobs and derivation parameters.
    /// </summary>
    /// <param name="vaultKey">The server's vault key response.</param>
    /// <returns>Task.</returns>
    public async Task CacheVaultKeyBlobsAsync(VaultKeyResponse vaultKey)
    {
        await localStorage.SetItemAsStringAsync(StorageKeys.EncryptedAccountKey, vaultKey.EncryptedAccountKey);
        if (!string.IsNullOrEmpty(vaultKey.EncryptedVek))
        {
            await localStorage.SetItemAsStringAsync(StorageKeys.EncryptedVek, vaultKey.EncryptedVek);
        }
        else
        {
            await localStorage.RemoveItemAsync(StorageKeys.EncryptedVek);
        }

        if (!string.IsNullOrEmpty(vaultKey.AccountPublicKey) && !string.IsNullOrEmpty(vaultKey.EncryptedAccountPrivateKey))
        {
            await localStorage.SetItemAsStringAsync(StorageKeys.AccountPublicKey, vaultKey.AccountPublicKey);
            await localStorage.SetItemAsStringAsync(StorageKeys.EncryptedAccountPrivateKey, vaultKey.EncryptedAccountPrivateKey);
        }
        else
        {
            await localStorage.RemoveItemsAsync([StorageKeys.AccountPublicKey, StorageKeys.EncryptedAccountPrivateKey]);
        }

        await StoreDerivationParamsAsync(new EncryptionKeyDerivationParams(vaultKey.Salt, vaultKey.EncryptionType, vaultKey.EncryptionSettings, true));
    }

    /// <summary>
    /// Remove the cached chain (the account has no chain, or the user logged out).
    /// </summary>
    /// <returns>Task.</returns>
    public async Task ClearCachedChainAsync()
    {
        await localStorage.RemoveItemsAsync([StorageKeys.EncryptedAccountKey, StorageKeys.EncryptedVek, StorageKeys.AccountPublicKey, StorageKeys.EncryptedAccountPrivateKey]);

        // Keep the flag in step with the chain, otherwise an offline unlock refuses to fall back to the legacy path.
        var parameters = await GetDerivationParamsAsync();
        if (parameters is { HasKeyChain: true })
        {
            await StoreDerivationParamsAsync(parameters with { HasKeyChain = false });
        }
    }

    /// <summary>
    /// Derive the KEK from a password.
    /// </summary>
    /// <param name="password">The password.</param>
    /// <param name="salt">The Argon2 salt.</param>
    /// <param name="encryptionSettings">The Argon2 settings JSON.</param>
    /// <returns>The derived key as base64.</returns>
    private async Task<string> DeriveKeyAsync(string password, string salt, string encryptionSettings)
    {
        return Convert.ToBase64String(await rustCoreService.Argon2DeriveKeyAsync(password, salt, encryptionSettings));
    }

    /// <summary>
    /// Decrypt the locally cached chain with the given KEK.
    /// </summary>
    /// <param name="derivedKeyBase64">The password-derived key.</param>
    /// <returns>The resolved keys, or null when no chain is cached.</returns>
    private async Task<ResolvedVaultKey?> ResolveFromLocalCacheAsync(string derivedKeyBase64)
    {
        var encryptedAccountKey = await localStorage.GetItemAsStringAsync(StorageKeys.EncryptedAccountKey);
        if (string.IsNullOrEmpty(encryptedAccountKey))
        {
            return null;
        }

        var encryptedVek = await localStorage.GetItemAsStringAsync(StorageKeys.EncryptedVek);
        var encryptedAccountPrivateKey = await localStorage.GetItemAsStringAsync(StorageKeys.EncryptedAccountPrivateKey);
        return await DecryptKeyChainAsync(encryptedAccountKey, encryptedVek, encryptedAccountPrivateKey, derivedKeyBase64);
    }

    /// <summary>
    /// Walk the chain: AK first, then the VEK (or the AK itself for a transitional account where AK equals VEK), then the account private key.
    /// </summary>
    /// <param name="encryptedAccountKey">The AK encrypted with the KEK.</param>
    /// <param name="encryptedVek">The VEK encrypted with the AK, or null for a transitional account.</param>
    /// <param name="encryptedAccountPrivateKey">The account private key encrypted with the AK, or null when the account has no keypair.</param>
    /// <param name="derivedKeyBase64">The password-derived KEK.</param>
    /// <returns>The resolved keys.</returns>
    private async Task<ResolvedVaultKey> DecryptKeyChainAsync(string encryptedAccountKey, string? encryptedVek, string? encryptedAccountPrivateKey, string derivedKeyBase64)
    {
        var accountKey = await DecryptKeyOrThrowAsync(encryptedAccountKey, derivedKeyBase64);
        var vek = string.IsNullOrEmpty(encryptedVek) ? accountKey : await DecryptKeyOrThrowAsync(encryptedVek, accountKey);
        return new ResolvedVaultKey(vek, await DecryptPrivateKeyAsync(encryptedAccountPrivateKey, accountKey), false);
    }

    /// <summary>
    /// Decrypt the account private key with the AK. A stale or corrupt blob must not fail the unlock; grant decryption then degrades until the next sync.
    /// </summary>
    /// <param name="encryptedAccountPrivateKey">The encrypted private key, or null.</param>
    /// <param name="accountKeyBase64">The Account Key.</param>
    /// <returns>The private key as a JWK JSON string, or null.</returns>
    private async Task<string?> DecryptPrivateKeyAsync(string? encryptedAccountPrivateKey, string accountKeyBase64)
    {
        if (string.IsNullOrEmpty(encryptedAccountPrivateKey))
        {
            return null;
        }

        try
        {
            return await jsInteropService.SymmetricDecrypt(encryptedAccountPrivateKey, accountKeyBase64);
        }
        catch (JSException ex)
        {
            logger.LogWarning(ex, "The account private key could not be decrypted; shared vault grants stay closed until the next sync.");
            return null;
        }
    }

    /// <summary>
    /// Decrypt a wrapped key, mapping an AES-GCM authentication failure onto <see cref="VaultKeyDecryptException"/>.
    /// </summary>
    /// <param name="encryptedKey">The wrapped key as base64(IV | ciphertext | tag).</param>
    /// <param name="decryptingKeyBase64">The key that wraps it.</param>
    /// <returns>The unwrapped key as base64.</returns>
    private async Task<string> DecryptKeyOrThrowAsync(string encryptedKey, string decryptingKeyBase64)
    {
        try
        {
            return Convert.ToBase64String(await jsInteropService.SymmetricDecryptBase64ToBytes(encryptedKey, decryptingKeyBase64));
        }
        catch (JSException ex)
        {
            throw new VaultKeyDecryptException(ex);
        }
    }
}

File: core/client/src/auth/VaultKeyService.ts
/**
 * VaultKeyService with client-side helpers for key operations.
 */

import { UnlockMethodType, type VaultKeyGetResponse, type VaultKeyResponse } from '@aliasvault/models/webapi';

import { ApiRequestError } from '../api/errors/ApiRequestError';
import { AppErrorCode, formatErrorWithCode } from '../api/errors/AppErrorCodes';
import { WebApiService } from '../api/WebApiService';
import { StorageKeys } from '../constants/StorageKeys';
import { EncryptionUtility } from '../crypto/EncryptionUtility';
import { getPlatform } from '../platform/ClientPlatform';

import type { UnlockKeyDerivationParams } from '@aliasvault/models/metadata';

/**
 * Result of fetching the vault key from the server.
 */
export type FetchVaultKeyResult = {
  supported: boolean;
  vaultKey: VaultKeyResponse | null;
};

/**
 * The keys an unlock key opens.
 */
export type SessionKeys = {
  /** The vault encryption key (VEK). */
  vaultEncryptionKey: string;
  /** The account private key (JWK). */
  accountPrivateKey: string | null;
};

/**
 * Static helper for fetching, caching and opening the account-key unlock chain (KEK → AK → VEK + account keypair).
 */
export class VaultKeyService {
  /**
   * Fetch the current user's password vault key from the server.
   * @param webApi - the API client to use (popup context passes its own instance; background creates one)
   */
  public static async fetchVaultKey(webApi?: WebApiService): Promise<FetchVaultKeyResult> {
    const api = webApi ?? new WebApiService();
    try {
      const response = await api.get<VaultKeyGetResponse>(`VaultKey/${UnlockMethodType.Password}`);
      return { supported: true, vaultKey: response.vaultKey ?? null };
    } catch (e) {
      if (e instanceof ApiRequestError && e.statusCode === 404) {
        return { supported: false, vaultKey: null };
      } else if (e instanceof Error && e.message.includes('status: 404')) {
        return { supported: false, vaultKey: null };
      }
      throw e;
    }
  }

  /**
   * Right after authentication: fetch the account's key chain from the server, check that the unlock key opens it
   * and cache it for offline unlock.
   * @param unlockKeyBase64 - the password-derived key (the KEK)
   * @param webApi - the API client to use
   * @throws Error with {@link AppErrorCode.VAULT_DECRYPT_FAILED} when the key does not open the chain (wrong password).
   */
  public static async refreshKeyChain(unlockKeyBase64: string, webApi?: WebApiService): Promise<void> {
    const result = await VaultKeyService.fetchVaultKey(webApi);

    if (!result.supported) {
      // Older server: trust the local cache.
      await VaultKeyService.verifyUnlockKey(unlockKeyBase64);
      return;
    }

    if (!result.vaultKey) {
      await getPlatform().storage.removeMany([StorageKeys.ENCRYPTED_VEK, StorageKeys.ENCRYPTED_ACCOUNT_KEY, StorageKeys.ACCOUNT_PUBLIC_KEY, StorageKeys.ENCRYPTED_ACCOUNT_PRIVATE_KEY]);
      return;
    }

    await VaultKeyService.openChain(unlockKeyBase64, result.vaultKey.encryptedAccountKey, result.vaultKey.encryptedVek ?? null, null);
    await VaultKeyService.cacheVaultKeyBlobs(result.vaultKey);
  }

  /**
   * Check offline that the unlock key opens the locally cached chain. A legacy account has no chain to check against.
   * @param unlockKeyBase64 - the password-derived key (the KEK), typed in or restored by PIN
   * @throws Error with {@link AppErrorCode.VAULT_DECRYPT_FAILED} when the key does not open the chain (wrong password).
   */
  public static async verifyUnlockKey(unlockKeyBase64: string): Promise<void> {
    await VaultKeyService.openKeyChain(unlockKeyBase64);
  }

  /**
   * Whether this device holds a key chain, i.e. whether the account is on the account-key model rather than a
   * legacy account whose unlock key encrypts the vault directly. The cache is written on every login and cleared
   * when the server reports no vault key, so it needs no server round-trip. A false answer is only ever stale in
   * one direction (another device migrated since the last login), which the sync resolves by adopting the remote chain.
   */
  public static async hasLocalVaultKey(): Promise<boolean> {
    return (await getPlatform().storage.get(StorageKeys.ENCRYPTED_ACCOUNT_KEY) as string | null) !== null;
  }

  /**
   * The unlock key of this session (the password-derived KEK), or null when the vault is locked.
   */
  public static async getSessionUnlockKey(): Promise<string | null> {
    return (await getPlatform().storage.get(StorageKeys.UNLOCK_KEY)) as string | null;
  }

  /**
   * The keys of the unlocked session, derived from the unlock key and the cached chain, or null when the vault is locked.
   */
  public static async getSessionKeys(): Promise<SessionKeys | null> {
    const unlockKey = await VaultKeyService.getSessionUnlockKey();
    return unlockKey ? VaultKeyService.openKeyChain(unlockKey) : null;
  }

  /**
   * The vault encryption key of the unlocked session, or null when the vault is locked.
   */
  public static async getSessionVaultEncryptionKey(): Promise<string | null> {
    return (await VaultKeyService.getSessionKeys())?.vaultEncryptionKey ?? null;
  }

  /**
   * The account private key of the unlocked session (JWK string), or null when the vault is locked or
   * the account has no keypair yet (legacy account, not migrated to manifest-v1 yet). Used to decrypt shared-manifest VEK grants.
   */
  public static async getSessionAccountPrivateKey(): Promise<string | null> {
    return (await VaultKeyService.getSessionKeys())?.accountPrivateKey ?? null;
  }

  /**
   * The cached account public key, or null when the account has no keypair yet.
   */
  public static async getAccountPublicKey(): Promise<string | null> {
    return (await getPlatform().storage.get(StorageKeys.ACCOUNT_PUBLIC_KEY)) as string | null;
  }

  /**
   * Persist new account key after a local password change. The session moves onto the new unlock key with it.
   * @param newEncryptedAccountKey - the Account Key encrypted with the new password-derived KEK
   * @param derivationParams - the KEK derivation parameters of the new password
   * @param newUnlockKeyBase64 - the new password-derived KEK
   */
  public static async persistNewAccountKey(newEncryptedAccountKey: string, derivationParams: UnlockKeyDerivationParams, newUnlockKeyBase64: string): Promise<void> {
    await getPlatform().storage.setMany([
      { key: StorageKeys.ENCRYPTED_ACCOUNT_KEY, value: newEncryptedAccountKey },
      { key: StorageKeys.UNLOCK_KEY_DERIVATION_PARAMS, value: derivationParams },
      { key: StorageKeys.UNLOCK_KEY, value: newUnlockKeyBase64 },
    ]);
  }

  /**
   * Open the locally cached chain with the unlock key. Without a cached chain (legacy account) the unlock key is
   * the vault encryption key.
   * @param unlockKeyBase64 - the password-derived key (the KEK)
   * @throws Error with {@link AppErrorCode.VAULT_DECRYPT_FAILED} when the key does not open the chain.
   */
  public static async openKeyChain(unlockKeyBase64: string): Promise<SessionKeys> {
    const storage = getPlatform().storage;
    const [encryptedAccountKey, encryptedVek, encryptedAccountPrivateKey] = await Promise.all([
      storage.get<string>(StorageKeys.ENCRYPTED_ACCOUNT_KEY),
      storage.get<string>(StorageKeys.ENCRYPTED_VEK),
      storage.get<string>(StorageKeys.ENCRYPTED_ACCOUNT_PRIVATE_KEY),
    ]);
    if (!encryptedAccountKey) {
      return { vaultEncryptionKey: unlockKeyBase64, accountPrivateKey: null };
    }
    return VaultKeyService.openChain(unlockKeyBase64, encryptedAccountKey, encryptedVek, encryptedAccountPrivateKey);
  }

  /**
   * Walk a chain: the unlock key decrypts the Account Key, which decrypts the VEK and the account private key.
   * @param unlockKeyBase64 - the password-derived key (the KEK)
   * @param encryptedAccountKey - the Account Key encrypted with the KEK
   * @param encryptedVek - the VEK encrypted with the Account Key
   * @param encryptedAccountPrivateKey - the account private key encrypted with the Account Key, or null when the account has none yet
   */
  private static async openChain(unlockKeyBase64: string, encryptedAccountKey: string, encryptedVek: string | null, encryptedAccountPrivateKey: string | null): Promise<SessionKeys> {
    if (!encryptedVek) {
      throw new Error('Vault key chain is missing the encrypted VEK');
    }

    const accountKey = await VaultKeyService.decryptKeyOrThrow(encryptedAccountKey, unlockKeyBase64);
    const vaultEncryptionKey = await VaultKeyService.decryptKeyOrThrow(encryptedVek, accountKey);

    let accountPrivateKey: string | null = null;
    if (encryptedAccountPrivateKey) {
      try {
        accountPrivateKey = await EncryptionUtility.symmetricDecrypt(encryptedAccountPrivateKey, accountKey);
      } catch {
        // A stale/corrupt private-key blob must not fail the unlock; grant decryption degrades until the next login.
      }
    }
    return { vaultEncryptionKey, accountPrivateKey };
  }

  /**
   * Persist a server vault-key response's encrypted blobs for offline unlock.
   * @param vaultKey - the server's vault key response
   */
  public static async cacheVaultKeyBlobs(vaultKey: VaultKeyResponse): Promise<void> {
    await getPlatform().storage.setMany([
      { key: StorageKeys.ENCRYPTED_ACCOUNT_KEY, value: vaultKey.encryptedAccountKey },
      { key: StorageKeys.ENCRYPTED_VEK, value: vaultKey.encryptedVek },
    ]);

    if (vaultKey.accountPublicKey && vaultKey.encryptedAccountPrivateKey) {
      await getPlatform().storage.set(StorageKeys.ACCOUNT_PUBLIC_KEY, vaultKey.accountPublicKey);
      await getPlatform().storage.set(StorageKeys.ENCRYPTED_ACCOUNT_PRIVATE_KEY, vaultKey.encryptedAccountPrivateKey);
    } else {
      await getPlatform().storage.removeMany([StorageKeys.ACCOUNT_PUBLIC_KEY, StorageKeys.ENCRYPTED_ACCOUNT_PRIVATE_KEY]);
    }
  }

  /**
   * Decrypt an encrypted key blob, mapping an AES-GCM authentication failure onto the standard decrypt-failed
   * error code so existing wrong-password handling applies.
   * @param encryptedKey - encrypted key blob
   * @param decryptingKeyBase64 - the key that decrypts it (the KEK derived from the unlock method)
   */
  private static async decryptKeyOrThrow(encryptedKey: string, decryptingKeyBase64: string): Promise<string> {
    try {
      return await EncryptionUtility.decryptVaultEncryptionKey(encryptedKey, decryptingKeyBase64);
    } catch {
      // E-203: decrypt failed, which for the password key type means the entered password is wrong.
      throw new Error(formatErrorWithCode('Failed to decrypt vault encryption key', AppErrorCode.VAULT_DECRYPT_FAILED));
    }
  }
}

export default VaultKeyService;
